File: src/plan.ts
import { PlanValidationError } from "./errors.js"
import { StepType, type Plan } from "./models.js"

export function validatePlan(plan: Plan): string[] {
  const errors: string[] = []

  if (!plan.planId) {
    errors.push("planId is required")
  }
  if (!plan.taskId) {
    errors.push("taskId is required")
  }
  if (!plan.goal.trim()) {
    errors.push("goal is required")
  }

  if (plan.constraints.allowedPaths.length === 0) {
    errors.push("constraints.allowedPaths must not be empty")
  }
  if (plan.constraints.maxFilesChanged <= 0) {
    errors.push("constraints.maxFilesChanged must be > 0")
  }

  if (plan.steps.length === 0) {
    errors.push("at least one plan step is required")
    return errors
  }

  const seenStepIds = new Set<string>()
  let hasCode = false
  let hasTest = false

  for (const step of plan.steps) {
    if (!step.id) {
      errors.push("every step must have an id")
    } else if (seenStepIds.has(step.id)) {
      errors.push(`duplicate step id: ${step.id}`)
    } else {
      seenStepIds.add(step.id)
    }

    if (step.type === StepType.CODE) {
      hasCode = true
    }
    if (step.type === StepType.TEST) {
      hasTest = true
      if (!step.command || !step.command.trim()) {
        errors.push(`test step '${step.id}' requires command`)
      }
    }
  }

  if (!hasCode) {
    errors.push("plan requires at least one code step")
  }
  if (!hasTest) {
    errors.push("plan requires at least one test step")
  }

  return errors
}

export function assertPlanValid(plan: Plan): void {
  const errors = validatePlan(plan)
  if (errors.length > 0) {
    throw new PlanValidationError(errors.join("; "))
  }
}

File: src/lucy_orchestrator/plan.py
from __future__ import annotations

from .exceptions import PlanValidationError
from .models import Plan, StepType


def validate_plan(plan: Plan) -> list[str]:
    errors: list[str] = []

    if not plan.plan_id:
        errors.append("plan_id is required")
    if not plan.task_id:
        errors.append("task_id is required")
    if not plan.goal.strip():
        errors.append("goal is required")

    if not plan.constraints.allowed_paths:
        errors.append("constraints.allowed_paths must not be empty")
    if plan.constraints.max_files_changed <= 0:
        errors.append("constraints.max_files_changed must be > 0")

    if not plan.steps:
        errors.append("at least one plan step is required")
        return errors

    seen_step_ids: set[str] = set()
    step_types: set[StepType] = set()

    for step in plan.steps:
        if not step.id:
            errors.append("every step must have an id")
        elif step.id in seen_step_ids:
            errors.append(f"duplicate step id: {step.id}")
        else:
            seen_step_ids.add(step.id)

        step_types.add(step.step_type)
        if step.step_type == StepType.TEST and not (
            step.command and step.command.strip()
        ):
            errors.append(f"test step '{step.id}' requires command")

    if StepType.CODE not in step_types:
        errors.append("plan requires at least one code step")
    if StepType.TEST not in step_types:
        errors.append("plan requires at least one test step")

    return errors


def assert_plan_valid(plan: Plan) -> None:
    errors = validate_plan(plan)
    if errors:
        raise PlanValidationError("; ".join(errors))
